### ISSUE TITLE
Propagate type of JSX elements up to parent elements 

### ISSUE THREAD

**Issue by stefan-wullems** [microsoft/TypeScript#40870]
## Suggestion

Propagate type of JSX elements up to parent elements so a parent element can infer its type based on the type of its children.

## Use Cases

I want to be able to compose React components to achieve a certain goal with the same type safety as I would have if I had if I had only one component that achieves that same goals but does so entirely through props.

## Examples

As an example, look at this tabs component in React.

```tsx
import React from 'react'

interface TabProps<N extends string> {
  name: N
}

function Tab<N extends string>(props: TabProps<N>) {
  return null
}

type TabsType<N extends string = string> =
  | React.FunctionComponentElement<TabProps<N>>
  | React.FunctionComponentElement<TabProps<N>>[]

interface TabsProps<C extends TabsType> {
  children: C
  active: C extends TabsType<infer N> ? N : unknown
}

export function Tabs<C extends TabsType>(props: TabsProps<C>) {
  return null
}

const test = (
  {/* I want active to be inferred as type 'bla' | 'hey' */}
  <Tabs active="blabla"> 
    <Tab name="bla" />
    <Tab name="hey" />
  </Tabs>
)
```

The reason why this doesn't work is that that the `C` type parameter of the `Tabs` component gets inferred as `JSX.Element` instead of `React.FunctionComponentElement<TabProps<N>>[]` which has the consequence that the default type: `string` gets used instead, making the invalid value blabla go unnoticed by typescript.

I could fix this by defining tabs like this:

```tsx
import React from 'react'

interface Tab {
   name: string
   content: string
}

interface TabsProps<T extends Tab[]> {
  tabs: T
  active: D[number]['name']
}

function Tabs<T extends Tab[]>(props: TabsProps<T>) {
  return null
}

 const tabs = [
  {  name: 'bla', content: 'one' },
  {  name: 'hey', content: 'two' }
] as const

const test = (
  <Tabs data={tabs} active="bla" />
)
```

But this is not the react way of approaching this problem. Also, with dynamic data it would lead to the need to memoize to prevent potentially expensive rerenders. This memoization can really clutter up your code a lot.

It feels like I have to choose between complete type safety, or elegant react patterns, but both at the same time are often not an option. Parents elements being able to infer the types of its children would solve some of these problems.

## Checklist

My suggestion meets these guidelines:

* [ ] This wouldn't be a breaking change in existing TypeScript/JavaScript code
* [x] This wouldn't change the runtime behavior of existing JavaScript code
* [x] This could be implemented without emitting different JS based on the types of the expressions
* [x] This isn't a runtime feature (e.g. library functionality, non-ECMAScript syntax with JavaScript output, etc.)
* [x] This feature would agree with the rest of [TypeScript's Design Goals](https://github.com/Microsoft/TypeScript/wiki/TypeScript-Design-Goals).

**Comment by stefan-wullems** [comment]
After doing some more research I found this issue: #14729. Am I correct that implementing that feature would fix this also?

**Comment by RyanCavanaugh** [comment]
Correct; #14729 is the canonical issue for tracking children (which in turn is only useful if the expression type of JSX elements is somehow differented)
